certstore_linux.go

```diff
@@ -1,14 +1,320 @@
+		certs      = C.PK11_ListCerts(C.PK11CertListUser, nil)
+		node       *C.CERTCertListNode
+	)
+	if certs == nil {
+		return nil, fmt.Errorf("error listing user certificates: %s", C.GoString(C.GetErrorString()))
+	}
+	defer C.CERT_DestroyCertList(certs)
+	for node = C.CertListHead(certs); C.CertListEnd(node, certs) == 0; node = C.CertListNext(node) {
+		identity := nssIdentity(*C.CERT_DupCertificate(node.cert))
+		identities = append(identities, &identity)
+	}
+	return identities, nil
+}
+
+// nssIdentity is a wrapper around a C.CERTCertificate.
+type nssIdentity C.CERTCertificate
+
+// Signer implements the Identity interface.
+func (i *nssIdentity) Signer() (crypto.Signer, error) {
+	return i, nil
+}
+
+// Certificate implements the Identity interface.
+func (i *nssIdentity) Certificate() (*x509.Certificate, error) {
+	var (
+		der   = i.derCert
+		bytes = C.GoBytes(unsafe.Pointer(der.data), C.int(der.len))
+	)
+	cert, err := x509.ParseCertificate(bytes)
+	if err != nil {
+		return nil, err
+	}
+	return cert, nil
+}
+
+// CertificateChain implements the Identity interface.
+func (i *nssIdentity) CertificateChain() ([]*x509.Certificate, error) {
+	var (
+		cert  = C.CERTCertificate(*i)
+		certs = C.CERT_GetCertChainFromCert(&cert, C.PR_Now(), C.certUsageAnyCA)
+	)
+	if certs == nil {
+		return nil, fmt.Errorf("error building certificate chain: %s", C.GoString(C.GetErrorString()))
+	}
+	defer C.CERT_DestroyCertList(certs)
+
+	var (
+		certificates = make([]*x509.Certificate, 0)
+		node         *C.CERTCertListNode
+	)
+	for node = C.CertListHead(certs); C.CertListEnd(node, certs) == 0; node = C.CertListNext(node) {
+		var (
+			identity  = nssIdentity(*node.cert)
+			cert, err = identity.Certificate()
+		)
+		if err != nil {
+			return nil, errors.New("unable to parse certificate when building chain")
+		}
+		certificates = append(certificates, cert)
+	}
+	return certificates, nil
 }
 
-// Implement this function, just to silence other compiler errors.
+// Delete implements the Identity interface.
+func (i *nssIdentity) Delete() error {
+	var (
+		cert      = C.CERTCertificate(*i)
+		secstatus = C.PK11_DeleteTokenCertAndKey(&cert, nil)
+	)
+	if secstatus != C.SECSuccess {
+		return errors.New("failed to delete certificate & key pair")
+	}
+	return nil
+}
+
+// Close implements the Identity interface.
+func (i *nssIdentity) Close() {
+	cert := C.CERTCertificate(*i)
+	C.CERT_DestroyCertificate(&cert)
+}
+
+// Public implements the crypto.Signer interface.
+func (i *nssIdentity) Public() crypto.PublicKey {
+	var cert, _ = i.Certificate()
+	if cert == nil {
+		return nil
+	}
+	return cert.PublicKey
+}
+
+// Sign implements the crypto.Signer interface.
+func (i *nssIdentity) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) ([]byte, error) {
+	hash := opts.HashFunc()
+	if len(digest) != hash.Size() {
+		return nil, errors.New("bad digest for hash")
+	}
+	var (
+		cert = C.CERTCertificate(*i)
+		key  = C.PK11_FindKeyByAnyCert(&cert, nil)
```

Comment:
The docs for `PK11_FindKeyByAnyCert` say:

>  When you are finished with the private key structure returned by PK11_FindKeyByAnyCert, you must free it by calling SECKEY_DestroyPrivateKey.

but I don't see a call to SECKEy_DestroyPrivateKey here?  It'd probably be best to do it in a `defer` immediately after we've confirmed we didn't get a `nil` response.

Reply:
Good catch. Fixed.